Audit workflows

teresaromero opened the pull request:
Fixes #2872 

This PR proposes some changes in order to improve the workflows when opening a PR. Ideally:
- developer runs `make lint` locally prior to commit&push code - this will surface pre-commit and golangci-lint issues
- once PR is created, lint workflow will run pre-commit
- once PR is created, deepsource will perform static analysis
- code coverage is availabe at Codecov.io after circle-ci unit test job has finished
- merge to master is protected by commit signature verification
- merge to master is protected by lint workflow success
- merge to master is protected by circle-ci workflows success

## Proposed changes

### Static Analysis

- updated DeepSource configuration

TODO:
- [x] Verify Analysis is set to Granular, if not update [Analysis settings to Granular](https://deepsource.io/docs/dashboard/repo-settings#analysis-settings) EDIT: Deepsource already has granular analysis
- [ ] Follow up Issue fixing all Deepsource errors for all code so each PR is clean

### Linters

- updated `make lint` to use pre-commit and golangci-lint
- updated `contribution.md` to describe how linting works at the repo
- removed super-lintern in favor of shell pre-commit hook
- removed trailing-whitespaces at workflow/lint.yml in favor of pre-commit hook
- removed misspell at workflow/link.yml in favor of codespell at pre-commit hook
- updated codespell pre-commit hook params and deleted codespell.txt & scripts related
- moved markdown and yaml lint configs to root and updated pre-commit hooks used. Some issues came around markdown files, so `bug_report`, `how-does-it-work`, and `pull_request_template` have some minor changes
- updated `.golangci.yaml` to run all linterns, have disabled for now the ones that we where hitting some error, as this could be an opportunity to address the errors or leave the exceptions. From my perspective, these disabled linters should be a follow-up issue to try to remove from that list as much as we can/want.

TODO:
- [ ] make workflow/lint.yml required to merge to master. Add a protection under the repository settings for master branch
- [ ] Follow-up issues fixing linting problems found by golang-ci

### Test coverage

- added .codecov.yml with settings to show just project and patch status as informational under workflows and avoid comments and annotations. This way the developer can access if needed the information at codecov but this is not a stopper to merge


### Contribution

- keep DEREK.yml from repository for verify commits control
- enforce commit signature at PR review stage, in the future setting up this branch protection policy for master

TODO:
- [ ] <del>Set the option to require commit signature verification under master branch protection<del> This will be done in the future

### Dependencies check

- no changes

jLopezbarb commented on .github/workflows/lint.yml:
```diff
@@ -1,44 +1,13 @@
-      - uses: actions/checkout@v3
-      - name: Check for trailing whitespace
-        run: "! git grep -EIn $'[ \t]+$' | grep -v '\\- '"
+      - uses: actions/setup-python@v3
```
why are we downgrading this?

teresaromero replied:
now we are using the action to run the pre-commit, instead of installing it and running the command on a custom action 
https://github.com/marketplace/actions/pre-commit

the action is stated to be deprecated for new features but for what we are needing here i think is enough to run the checks. WDYT?

jLopezbarb replied:
I think we should use the latest version of the action and pin it to an specific python version. That way we control which python version we are executing and can troubleshoot any problem

teresaromero replied:
yes, i reverted this... the change was made beacuse i changed to use the pre-commit action instead of the installation at the action, and at the example they are using v3

cc: @ifbyol

jLopezbarb commented on .golangci.yml:
```diff
@@ -1,3 +1,47 @@
+    - nosprintfhostport
+    - gosec
     - errcheck
+  fast: true
```
Should we run without this flag?

teresaromero replied:
yes, i've revised the docs and updated the file

jLopezbarb commented on .pre-commit-config.yaml:
```diff
@@ -20,56 +22,44 @@ repos:
   - repo: https://github.com/adrienverge/yamllint
     rev: v1.26.3
     hooks:
       - id: yamllint
```
Isn't this covered by markdownlint in line 38: https://github.com/okteto/okteto/pull/3036/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R38

jLopezbarb commented on .codecov.yml:
```diff
@@ -0,0 +1,13 @@
+
+comment: false
+github_checks:
+  annotations: false
```
Thanks for removing the annotations on the PRs. That make it difficult to follow a review!!❤️

AdrianPedriza commented on .mdlrc:
```diff
@@ -1,24 +1,8 @@
-# https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md
-# markdownlint -c .github/linters/.markdown-lint.yml .
-
+rules '~MD013', '~MD024', '~MD025', '~MD026', '~MD033', '~MD040', '~MD041'
```
it is just a refactor of this file? I think that grouping rules+description as before may be clearer

teresaromero replied:
i've changed the pre-commit hook that checks this markdownlint, so it has different way to inform the rules

ifbyol commented on .DEREK.yml:
```diff
@@ -1,13 +0,0 @@
-  - jLopezbarb
-
-features:
-  - dco_check
```
Maybe we should keep this for some time instead of forcing the protection rule. Before enabling the protection rule I would like to check that automatic bots that sends PRs to the repo use verified commits. For example, would the backport flow keep working once we merge and configure the protection rule? Is that bot creating a signed commit?

ifbyol commented on .github/workflows/lint.yml:
```diff
@@ -1,44 +1,13 @@
-      - uses: actions/checkout@v3
-      - name: Check for trailing whitespace
-        run: "! git grep -EIn $'[ \t]+$' | grep -v '\\- '"
+      - uses: actions/setup-python@v3
```
Is it on purpose to downgrade the `setup-python` version? It seems we were using the version 4 and now the version 3

teresaromero replied:
yes, i've reverted this as per comment of javi

ifbyol commented on .typo-ci.yml:
```diff
@@ -1,17 +0,0 @@
-dictionaries:
```
Just curiosity, are we just removing it or is it replaced by something else?

teresaromero replied:
this file was not being used by any pipeline... so removed it. Is it part of any other workflow i am not aware of?

ifbyol replied:
I think I had some report of this in the past, but not sure where/how. Anyways, this is the configuration for [typo CI](https://typoci.com/documentation) and we don't have it configure, so it can be removed

teresaromero commented:
@ifbyol the DEREK comment has dissapear because i reverted .DEREK.yml i think.

About the concerns of bots:
- dependantbot creates verified commits, so in this case, any PR opened to master from this bot will comply with the branch protection
- backports are created agains release branches, so in this case, the github-actions bot does not create signed commits, but this would be a cherry-pick from a commit that has been created as part of a PR to master. 

We can keep DEREK just for the commit verification for a while, comments and description i've removed because i think we dont use this option to manage issues

AgustinRamiroDiaz commented on .golangci.yml:
```diff
@@ -1,3 +1,107 @@
+    # TODO: enable when fixed
+    - whitespace
+    - staticcheck
+# Enabled by your configuration linters:
```
Why do we have this comments here?

teresaromero replied:
this are the linters enabled by our configurations. I thought it would be interesting to have them here as a comment to know which ones are being applied. 

the TODO ones i disabled them as there where too many on code and this would need some PRs to clean the errors and enable them again

ifbyol commented:
> 

yeah, makes sense. I prefer to keep Derek for the sign verification, so we can start asking people to sign all the commits and after some time, we can enforce it at github protection level.

Can we update the PR description to match this change?

> enforce commit signature by setting up this branch protection policy for master

@teresaromero Do you need access to something to configure things?

teresaromero commented:
> > 
> 
> yeah, makes sense. I prefer to keep Derek for the sign verification, so we can start asking people to sign all the commits and after some time, we can enforce it at github protection level.
> 
> Can we update the PR description to match this change?

done! 

 
> @teresaromero Do you need access to something to configure things?

i only need the admin for deepsource to stablish the granular check, as in my dashboard i dont have access to settings
